modules/seqexec/server/src/main/scala/seqexec/server/ghost/GHOSTController.scala

```diff
@@ -152,48 +116,179 @@ object GHOSTController {
+                                   override val ifu1Coordinates: Coordinates,
+                                   ifu2Coordinates: Coordinates) extends HighResolutionMode {
+      override def ifu2Config: Configuration =
+        ifuConfig(IFUNum.IFU2, IFUTargetType.SkyPosition, ifu2Coordinates, BundleConfig.Sky)
+    }
+  }
+
+  // These are the parameters passed to GHOST from the WDBA.
+  // We use them to determine the type of configuration being used by GHOST, and instantiate it.
   object GHOSTConfig {
-    private implicit val durationEq: Eq[Duration] = Eq.by(_.toMillis)
-    implicit val eq: Eq[GHOSTConfig] = Eq.by(
-      x =>
-        (x.baseCoords,
-          x.expTime,
-          x.srifu1Name,
-          x.srifu1Coords,
-          x.srifu2Name,
-          x.srifu2Coords,
-          x.hrifu1Name,
-          x.hrifu1Coords,
-          x.hrifu2Name,
-          x.hrifu2Coords))
+    def apply(baseCoords: Option[Coordinates],
+              expTime: Duration,
+              srifu1Name: Option[String],
+              srifu1Coords: Option[Coordinates],
+              srifu2Name: Option[String],
+              srifu2Coords: Option[Coordinates],
+              hrifu1Name: Option[String],
+              hrifu1Coords: Option[Coordinates],
+              hrifu2Name: Option[String],
+              hrifu2Coords: Option[Coordinates]): Either[ExtractFailure, GHOSTConfig] = {
+      import IFUTargetType._
+
+      val sifu1 = determineType(srifu1Name)
+      val sifu2 = determineType(srifu2Name)
+      val hifu1 = determineType(hrifu1Name)
+      val hifu2 = determineType(hrifu2Name)
+      val extracted = (sifu1, sifu2, hifu1, hifu2) match {
+        case (Target, NoTarget, NoTarget, NoTarget) =>
+          srifu1Coords.map(StandardResolutionMode.SingleTarget.apply(baseCoords, expTime, _))
+        case (Target, Target, NoTarget, NoTarget) =>
+          (srifu1Coords, srifu2Coords).mapN(StandardResolutionMode.DualTarget.apply(baseCoords, expTime, _, _))
+        case (Target, SkyPosition, NoTarget, NoTarget) =>
+          (srifu1Coords, srifu2Coords).mapN(StandardResolutionMode.TargetPlusSky.apply(baseCoords, expTime, _, _))
+        case (SkyPosition, Target, NoTarget, NoTarget) =>
+          (srifu1Coords, srifu2Coords).mapN(StandardResolutionMode.SkyPlusTarget.apply(baseCoords, expTime, _, _))
+        case (NoTarget, NoTarget, Target, NoTarget) =>
+          hrifu1Coords.map(HighResolutionMode.SingleTarget.apply(baseCoords, expTime, _))
+        case (NoTarget, NoTarget, Target, SkyPosition) =>
+          (hrifu1Coords, hrifu2Coords).mapN(HighResolutionMode.TargetPlusSky.apply(baseCoords, expTime, _, _))
+        case _ =>
+          None
+      }
+      extracted match {
+        case Some(config) => Right(config)
+        case None         => Left(ContentError("Response does not constitute a valid GHOST configuration"))
+      }
+    }
 
+    implicit val eq: Eq[GHOSTConfig] = Eq.fromUniversalEquals
```

Comment:
Should this by `Eq.instance` and pattern match on the subclasses?

Reply:
I'll give tat a try. I'm not familiar with `Eq.instance` but it'll be a good learning experience.